### PR TITLE
#1650 - upgrades the dataloader version AND fixes lazy data loader support

### PR DESCRIPTION
**Pull request by bbakerman**
This upgrades to the latest version of java-dataloader which has lazy data loader support.

#1650 shows a problem when the data loaders in action are not present at the time the instrumentation gets started.

We now use a magic variable to know if NO data loader registry has been set by the consumer.

This allows us to retain the optimisation whereby we do no work when no data loaders are in effect

### CODE REVIEW

File: src/main/java/graphql/schema/DataFetchingEnvironmentImpl.java [hunk 1]
```diff
@@ -166,6 +166,11 @@ public ExecutionStepInfo getExecutionStepInfo() {
         return dataLoaderRegistry.getDataLoader(dataLoaderName);
     }
 
+    @Override
+    public DataLoaderRegistry getDataLoaderRegistry() {
+        return dataLoaderRegistry;
+    }
+
```
Comment (bbakerman) [inline src/main/java/graphql/schema/DataFetchingEnvironmentImpl.java]: I did this because we want to eventually get rid of ExecutionContext inside the DFE - so lets just get straight to the objects instead

File: src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java [hunk 2]
```diff
@@ -9,6 +12,14 @@
  */
 public class DataLoaderDispatcherInstrumentationState implements InstrumentationState {
 
+    @Internal
+    public static final DataLoaderRegistry EMPTY_DATALOADER_REGISTRY = new DataLoaderRegistry() {
+        @Override
+        public DataLoaderRegistry register(String key, DataLoader<?, ?> dataLoader) {
+            return Assert.assertShouldNeverHappen("You MUST set in your own DataLoaderRegistry to use data loader");
+        }
+    };
```
Comment (bbakerman) [inline src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java]: Today a person could technically take the default DLR from the execution input and use to to register in dataloaders.

But I really want to retain the optimisation thatw e do no work when they have not set a DLR.

I figure we have 2 choices.  Use a null in the ExecutionInput and null check or use a default well know one and blow up.

I can see arguments for either way.  Thoughts?

Comment (tinnou) [reply src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java]: > But I really want to retain the optimisation thatw e do no work when they have not set a DLR.

So if I understand correctly, you are trying to optimize only for apps that do not use Dataloader **at all**. 
You are not trying to optimize for apps that have a mix of queries where some leverage data loader and others don't? (although I think it would be hard to know ahead of time which queries would trigger data loading vs not)

--
Also, the default seems reasonable to avoid NPEs downstream.

File: src/main/java/graphql/schema/DataFetchingEnvironmentImpl.java (hunk 1 above)
Comment (tinnou) [reply src/main/java/graphql/schema/DataFetchingEnvironmentImpl.java]: Interesting, what's the purpose to getting rid of the ExecutionContext?

File: src/test/groovy/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationTest.groovy [hunk 3]
```diff
@@ -225,4 +230,40 @@ class DataLoaderDispatcherInstrumentationTest extends Specification {
         starWarsWiring.batchFunctionLoadCount == 2
         starWarsWiring.naiveLoadCount == 8
     }
+
+    def "can be efficient with lazily computed data loaders"() {
```
Comment (tinnou) [inline src/test/groovy/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationTest.groovy]: maybe we could add a negative test, where we attempt to register a dataloader onto the default registry and expect it to blow up.

File: src/main/java/graphql/schema/DataFetchingEnvironmentImpl.java (hunk 1 above)
Comment (bbakerman) [reply src/main/java/graphql/schema/DataFetchingEnvironmentImpl.java]: This is half wish in that the ExecutionContext really should have been an @Internal class.

For example it contans the "ExecutionStrategy"s to be used.  But that is not something that call back code like DataFecters should care about

It currently contains

```
    private final GraphQLSchema graphQLSchema;
    private final ExecutionId executionId;
    private final InstrumentationState instrumentationState;
    private final ExecutionStrategy queryStrategy;
    private final ExecutionStrategy mutationStrategy;
    private final ExecutionStrategy subscriptionStrategy;
    private final Map<String, FragmentDefinition> fragmentsByName;
    private final OperationDefinition operationDefinition;
    private final Document document;
    private final Map<String, Object> variables;
    private final Object root;
    private final Object context;
    private final Instrumentation instrumentation;
    private final List<GraphQLError> errors = new CopyOnWriteArrayList<>();
    private final Set<ExecutionPath> errorPaths = new HashSet<>();
    private final DataLoaderRegistry dataLoaderRegistry;
    private final CacheControl cacheControl;
    private final DeferSupport deferSupport = new DeferSupport();

```

There are useful things there but we regret making it all available as API

File: src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java (hunk 2 above)
Comment (bbakerman) [reply src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java]: We are trying to optimize for both ease of use AND performance

So We automatically add `DataLoaderDispatcherInstrumentation` to the mix if its not there

```
        boolean containsDLInstrumentation = instrumentationList.stream().anyMatch(instr -> instr instanceof DataLoaderDispatcherInstrumentation);
        if (!containsDLInstrumentation) {
            instrumentationList.add(new DataLoaderDispatcherInstrumentation());
        }
```

However if you dont use any DataLoaders then this can become a noop like thing so we cant be faster and no track all your fields and call dispatch on an empty registry.

if you look at `graphql.execution.instrumentation.dataloader.FieldLevelTrackingApproach` which is all the state we need to know what fields have completed and so on its non trivial

If your not using DataLoaders then you save memory and a little bit of CPU

Perhaps we should never have auto added `DataLoaderDispatcherInstrumentation` but we kept getting support request from the field saying data loaders dont work and we had to say "yeah you need to add a specvial `DataLoaderDispatcherInstrumentation` to make this work.

File: src/main/java/graphql/schema/DataFetchingEnvironmentImpl.java (hunk 1 above)
Comment (tinnou) [reply src/main/java/graphql/schema/DataFetchingEnvironmentImpl.java]: I see, add the useful things directly now makes more sense vs opening everything.

File: src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java (hunk 2 above)
Comment (tinnou) [reply src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java]: > Perhaps we should never have auto added DataLoaderDispatcherInstrumentation but we kept getting support request from the field saying data loaders dont work and we had to say "yeah you need to add a specvial DataLoaderDispatcherInstrumentation to make this work.

Yea and in that case we could have validated at the time of execution of a request that a `DataLoaderDispatcherInstrumentation` must be present if an ExecutionInput specified a `DataLoaderRegistry`. That way people would get an error and just add the instrumentation right away instead of opening an issue. Ease of use (auto-configuration) vs performance is definitely a balance though.

File: src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java [hunk 4]
```diff
@@ -20,8 +31,11 @@ public DataLoaderDispatcherInstrumentationState(Logger log, DataLoaderRegistry d
         this.dataLoaderRegistry = dataLoaderRegistry;
         this.approach = new FieldLevelTrackingApproach(log, dataLoaderRegistry);
         this.state = approach.createState();
-        hasNoDataLoaders = dataLoaderRegistry.getKeys().isEmpty();
-
+        //
```
Comment (JaidenAshmore) [inline src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java]: Are we actually get that much performance improvements by caching this `hasNoDataLoaders` field in the constructor compared to calling `dataLoaderRegistry.getKeys().isEmpty()` in the method `boolean hasNoDataLoaders()` considering `HashSet` for example just does:  `size == 0`.  Looking into the code the only significant negative would be the unnecessary cloning of the keys `HashSet` each time which we could improve by adding a new method like `DataLoaderRegistry#isEmpty` or `DataLoaderRegistry#hasLoaders`.  This would remove the need for cloning the keys `HashSet` when calling `dataLoaderRegistry#keys` and then we are literally just replacing:

```
if (state.hasNoDataLoaders)
```

with an equivalent

```
if (dataLoaderRegistry.dataLoaders.size == 0)
```

The advantage over this approach is that the consumers of the library do not need to care about this `EMPTY_DATALOADER_REGISTRY`, they can just add the `DataLoaderRegistry` that they like and it should work as expected for this use case or they can register no registry and it would still work as they intended.

The only negative to this approach I can see is whether we want to expose an `isEmpty` method onto the `DataLoaderRegistry` class? Maybe there is a reason for this that I am missing?

Comment (JaidenAshmore) [reply src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java]: Another alternative design decision could be to convert the `DataLoaderRegistry` to an interface or make an interface that it implements. When you provide none you include a `EmptyDataLoaderRegistry` with all your performance improvements needs for the most common case.

If the developer just wants a static list of data loaders they could use an `ImmutableDataLoaderRegistry` which is immutable and cannot change and will still get the performance needs.

If you need something like this use case where we want to be able to change the data loaders during execution the developer could construct it with a `MutableDataLoaderRegistry` so it can be dynamically changed during the runtime. This would not be as performant with these checks but would allow for this edge case.

File: src/test/groovy/graphql/util/TreeParallelTraverserTest.groovy [hunk 5]
```diff
@@ -26,19 +28,22 @@ class TreeParallelTraverserTest extends Specification {
...
+                        Awaitility.await().untilTrue(proceed)
                     }
                     nodes.add(number)
                     TraversalControl.CONTINUE
                 }
         ] as TraverserVisitor
         when:
-        TreeParallelTraverser.parallelTraverser({ n -> n.children }).traverse(root, visitor)
 
+        Timer timer = new Timer()
```
[the remark this answers is not in the document]
Comment (andimarek) [reply src/test/groovy/graphql/util/TreeParallelTraverserTest.groovy]: this is not releated to this PR
